Source/WebSocket/hcwebsocket.cpp

```diff
@@ -101,25 +109,71 @@ try
         return E_INVALIDARG;
     }
 
-    auto httpSingleton = get_http_singleton(true);
-    if (nullptr == httpSingleton)
-        return E_HC_NOT_INITIALISED;
+    std::shared_ptr<websocket_connect_context> connectContext = std::make_shared<websocket_connect_context>();
+    connectContext->uri = uri;
+    connectContext->subProtocol = subProtocol;
+    connectContext->websocket = websocket;
+    connectContext->outerAsyncBlock = asyncBlock;
+    websocket_connect_context* rawConnectContext = static_cast<websocket_connect_context*>(shared_ptr_cache::store<websocket_connect_context>(connectContext));
 
-    auto connectFunc = httpSingleton->m_websocketConnectFunc;
-    if (connectFunc != nullptr)
+    HRESULT hr = XAsyncBegin(asyncBlock, rawConnectContext, reinterpret_cast<void*>(HCWebSocketConnectAsync), __FUNCTION__,
+        [](_In_ XAsyncOp op, _In_ const XAsyncProviderData* data)
     {
-        try
+        switch (op)
         {
-            websocket->connectCalled = true;
-            connectFunc(uri, subProtocol, websocket, asyncBlock);
-        }
-        catch (...)
-        {
-            HC_TRACE_ERROR(WEBSOCKET, "HCWebSocketConnect [ID %llu]: failed", websocket->id);
+            case XAsyncOp::DoWork:
+            {
+                websocket_connect_context* rawConnectContext = static_cast<websocket_connect_context*>(data->context);
+                auto httpSingleton = get_http_singleton(true);
+                if (nullptr == httpSingleton)
+                    return E_HC_NOT_INITIALISED;
+
+                auto connectFunc = httpSingleton->m_websocketConnectFunc;
+                HRESULT hr = S_OK;
+                if (connectFunc != nullptr)
+                {
+                    try
+                    {
+                        rawConnectContext->websocket->connectCalled = true;
+                        hr = connectFunc(
+                            rawConnectContext->uri.c_str(),
+                            rawConnectContext->subProtocol.c_str(),
+                            rawConnectContext->websocket,
+                            rawConnectContext->outerAsyncBlock,
+                            httpSingleton->m_performEnv.get());
+                    }
+                    catch (...)
+                    {
+                        HC_TRACE_ERROR(WEBSOCKET, "HCWebSocketConnect [ID %llu]: failed", rawConnectContext->websocket->id);
+                    }
+                }
+                return E_PENDING;
+            }
+
+            case XAsyncOp::GetResult:
+                break;
+
+            case XAsyncOp::Cancel:
+                break;
+
+            case XAsyncOp::Cleanup:
+            {
+                websocket_connect_context* context = static_cast<websocket_connect_context*>(data->context);
+                //HCWebSocketCloseHandle(context->websocket); // TODO: jasonsa - fix for WinHTTP
```

Comment:
where does this get duplicated?

Reply:
bug.  it should be above.  the intent is that we keep the HCWebsocketHandle alive while connecting/connected, and let it be destroyed only when its disconnected.  we'll have to chat a bit about this